File: packages/python/python-shared/src/python_shared/http/pipeline/step_cache_eval.py
from python_shared.http.constants import HTTP_CONSTANTS
from python_shared.http.pipeline.types import PipelineContext

class StepCacheEval:
    name: str = "CacheEval"
    description: str = "Tenant Partitioned LRU Cache Lookup & Invalidation"

    async def execute(self, ctx: PipelineContext) -> None:
        ctx.step_index += 1
        method = ctx.config.method.upper()

        if method in (
            HTTP_CONSTANTS.METHOD_POST,
            HTTP_CONSTANTS.METHOD_PUT,
            HTTP_CONSTANTS.METHOD_PATCH,
            HTTP_CONSTANTS.METHOD_DELETE,
        ):
            ctx.cache_store.clear(ctx.tenant_id)
            return

        if method == HTTP_CONSTANTS.METHOD_GET and not ctx.config.no_cache and ctx.hashed_request_key:
            cached = ctx.cache_store.get(ctx.tenant_id, ctx.hashed_request_key)
            if cached is not None:
                ctx.cached_response = cached

File: packages/llm-obs-node-packages/shared-infra/src/http/pipeline/step-cache-eval.ts
/**
 * @file step-cache-eval.ts
 * @description Pipeline Step 6: Tenant LRU Cache Evaluation & Invalidation.
 */

import type { PipelineStep, PipelineContext } from "./types";
import { isCacheDisabled } from "../utils/http-utils";
import { HTTP_CONSTANTS } from "../constants";

export class StepCacheEval implements PipelineStep {
  public readonly name = "CacheEval";
  public readonly description = "Tenant Partitioned LRU Cache Lookup & Invalidation";

  public async execute<T>(ctx: PipelineContext<T>): Promise<void> {
    ctx.stepIndex++;
    const method = ctx.config.method.toUpperCase();

    if (([HTTP_CONSTANTS.METHOD_POST, HTTP_CONSTANTS.METHOD_PUT, HTTP_CONSTANTS.METHOD_PATCH, HTTP_CONSTANTS.METHOD_DELETE] as string[]).includes(method)) {

      ctx.cacheStore.clear(ctx.tenantId);
      return;
    }

    if (method === HTTP_CONSTANTS.METHOD_GET) {
      const disabled = isCacheDisabled(ctx.config.noCache, ctx.config.headers || {});
      if (!disabled && ctx.hashedRequestKey) {
        const cached = ctx.cacheStore.get<T>(ctx.tenantId, ctx.hashedRequestKey);
        if (cached !== undefined) {
          ctx.cachedResponse = cached;
          ctx.span?.setAttribute(HTTP_CONSTANTS.KEY_CACHE_HIT, true);
          ctx.span?.addEvent(HTTP_CONSTANTS.EVENT_CACHE_EVALUATED, {
            "cache.hit": true,
            "cache.key": ctx.hashedRequestKey,
          });
        }
      }
    }
  }
}
